Allow to export current reslices as image files

daniel-wer opened the pull request:
I've added a shortcut (`q`) for now to trigger the screenshot download. A button would probably be more discoverable. Maybe a small camera icon button at the top level (next to the position?!) or we put it in the dropdown menu :/

Getting an image from a data buffer was harder than I expected (and then webGL's y-axis is inverted to top it off ^^).

### URL of deployed dev instance (used for testing):
- https://screenshot.webknossos.xyz

### Steps to test:
- Press q in any view OR use the screenshot button in the dropdown in any tracing type. One (or more) save prompts should appear to download the currently visible viewport(s) as a png file.
- Names should contain the dataset name, current position and plane. (Add zoom?)
- Node selection should still work.

### Issues:
- fixes #3760

------
- [x] Updated [changelog](../blob/master/CHANGELOG.md#unreleased)
- [x] Ready for review

philippotto commented:
(While testing it was a bit weird to have three save dialogs pop up in ortho mode. Theoretically, this could be improved by bundling them in a zip. However, I think it's fine without, since it's probably not done that often.)

philippotto commented:
I just tested what happens when a viewport is hidden (tab inactive). In that case, an exception is thrown (due to width being zero). This should probably be checked for :)

I also noticed a subtle red (blue, green) border at the top and right edge of the screenshot. This is probably the clear color shining through? This is probably more an issue with placing the plane geometry correctly in the scene (so, independent of this PR) :S Just so that you are aware. I don't think it's blocking this PR. Might be related to #3785 ?

daniel-wer commented:
@philippotto Thanks for the thorough testing! I fixed the exception that was thrown when a viewport is hidden. I've also added the screenshot button. For that, I had to refactor the code a little to be able to access the screenshot functionality from the action bar. However, I think this refactoring has been very beneficial and I was able to deduplicate a lot of code in the process. Would be cool if you could have another look and do a last test round :)

I also saw the small colored borders, but don't know how to fix those, yet.

philippotto commented:
> Would be cool if you could have another look and do a last test round :)

Your refactoring looks really good :+1:

> I also saw the small colored borders, but don't know how to fix those, yet.

Due to the new feedback, I'm afraid we have to fix these before merging, right? Multiple thoughts I have:

- as a sanity check, I'd disable the css borders and double check whether the borders are in fact the clear color
- the border could be visible due to numeric imprecisions. maybe this is easy to fix by "rounding" appropriately at the right place?
- a dirty fix could be to simply make the plane a bit bigger than the viewport
- an even dirtier fix would be to crop the output image a bit further (2 pixels on each edge :see_no_evil:)

philippotto commented on frontend/javascripts/oxalis/view/arbitrary_view.js:
```diff
@@ -208,30 +212,6 @@ class ArbitraryView {
-
-    renderer.setViewport(0, 0, width, height);
-    renderer.setScissorTest(false);
-    renderer.setClearColor(0x222222, 1);
```
That particular clear color was problably not very important, right?

daniel-wer replied:
I had a look at the `getRenderedBucketsDebug` method which used `renderToTexture` before and couldn't find that the clear color would be important in any way for that.

daniel-wer commented:
Thanks for the debugging ideas!
The artifacts are indeed caused by the clear color, I confirmed by disabling the border.
I had a look at the code to position and scale the plane, but couldn't find any inaccuracies yet.
Enlarging the plane by 1 px does fix the artifacts (so numeric imprecisions seem very likely to be the cause), but I don't know whether this will have any other bad effects. Thinking about it, I wouldn't know why as the plane is still positioned properly, simply "continuing" 1px further to the right and bottom.

I'll invest some more time into finding the imprecision, but if I won't find anything, the slightly larger plane may be our best option.
